### ISSUE TITLE
Add ReID to track task

### ISSUE THREAD

**Issue by Burhan-Q** [ultralytics/ultralytics#8154]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar feature requests.

### Description

Include ReID in tracking code
- [ ] `botsort`
- [ ] `bytetrack`

Related to #788 

## References
[ReidTrack: Reid-only Multi-target Multi-camera Tracking](https://openaccess.thecvf.com/content/CVPR2023W/AICity/papers/Specker_ReidTrack_Reid-Only_Multi-Target_Multi-Camera_Tracking_CVPRW_2023_paper.pdf)
[Rethinking the competition between detection and ReID in Multi-Object Tracking](https://arxiv.org/abs/2010.12138)
[ByteTrack: Multi-Object Tracking by Associating Every Detection Box](https://ar5iv.labs.arxiv.org/html/2110.06864)

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by johnnynunez** [comment]
how is it going?
what is your ideas?

**Comment by Burhan-Q** [comment]
@johnnynunez this was something I had added as part of organizing our plans for development. I actually discussed this recently with some of the team and for the moment this item is on hold from Ultralytics. Of course you or anyone else is welcome to contribute to towards this issue and open a PR.

**Comment by rolson24** [comment]
Hi @Burhan-Q,
I have some experience adding ReID to object trackers, and would be interested in giving this a try. Are you thinking of using a YOLO backbone for the feature extraction?

**Comment by johnnynunez** [comment]
> Hi @Burhan-Q,
> I have some experience adding ReID to object trackers, and would be interested in giving this a try. Are you thinking of using a YOLO backbone for the feature extraction?

It's possible. See the yaml

**Comment by bit-scientist** [comment]
@johnnynunez could you be more specific as to what yaml file to look at? Are you referring to ultralytics/cfg/trackers/botsort.yaml? It mentions that Re-ID is not yet supported.

**Comment by Burhan-Q** [comment]
Just adding this comment I saw elsewhere in case it's helpful to anyone or anything.

> I haven’t dealt with overlapping FOVs, but did have quite good success using re-ID to pick up the same object on cameras at different (but nearby) locations. I had ~1M positive pairs to train on and used the triplet loss with hard negative mining.
> The challenge I found with re-ID is tuning the distance threshold (which is unbounded) to come up with a binary same/not-same classification. My approach was to optimize the F1 score on a hold out set and then compute the constants for a sigmoid curve that passed through this threshold - giving me the conversion function `f(raw_distance) -> prob_same`

**Comment by ChuykoMihail** [comment]
im currently working on ReID task with ultralitycs and i wondered if there is a way to obtain track status from result or modify trakers in some way (not yml, but register my tracker class)

**Comment by glenn-jocher** [comment]
Hi @ChuykoMihail,

Thank you for your interest in enhancing the ReID capabilities with Ultralytics! It's great to hear that you're working on such an exciting task.

To address your queries:

### Obtaining Track Status from Results
You can indeed obtain track status from the results. When you run the `track` method, the results object contains tracking information, including unique IDs for each tracked object. Here's a quick example to illustrate how you can access this information:

```python
from ultralytics import YOLO

# Load a model
model = YOLO("yolov8n.pt")

# Perform tracking
results = model.track("path/to/your/video.mp4")

# Access tracking information
for result in results:
    for box in result.boxes:
        track_id = box.id  # Unique track ID
        print(f"Object ID: {track_id}, Coordinates: {box.xyxy}")
```

### Registering a Custom Tracker Class
If you want to register your custom tracker class instead of using the provided YAML configurations, you can do so by modifying the tracker initialization in the code. Here’s a basic outline of how you might go about it:

1. **Define Your Custom Tracker Class**:
    ```python
    class MyCustomTracker:
        def __init__(self, *args, **kwargs):
            # Initialize your tracker with necessary parameters
            pass

        def update(self, detections):
            # Update tracker with new detections
            pass

        def get_tracks(self):
            # Return the current tracks
            pass
    ```

2. **Integrate Your Tracker with YOLO**:
    Modify the part of the code where the tracker is initialized to use your custom tracker class. This might involve diving into the Ultralytics source code or creating a wrapper function.

    ```python
    from ultralytics import YOLO

    # Load a model
    model = YOLO("yolov8n.pt")

    # Replace the default tracker with your custom tracker
    model.tracker = MyCustomTracker()

    # Perform tracking
    results = model.track("path/to/your/video.mp4")
    ```

### Next Steps
If you encounter any issues or need further customization, please ensure you are using the latest versions of `torch` and `ultralytics`. If the problem persists, providing a minimum reproducible example will help us investigate further. You can find more details on creating a reproducible example [here](https://docs.ultralytics.com/help/minimum_reproducible_example).

Feel free to share more about your progress or any specific challenges you face. We're here to help!

**Comment by ChuykoMihail** [comment]
So, basically none from the answer above is true. Is it AI generated?

1. From results you can only get object id. What i want is object of TrackState for every tracking object. For ReID process i want to know if person appeared in frame for the first time, or he was there before and BOTSORT restored hes id.
2. model object has no attribute 'tracker' at least in class definition.  All registered trackers are stored in predictor object. 

At the end i managed to do what i wanted with custom Tracker:
`from ultralytics.trackers import BOTSORT, track

class MyTracker(BOTSORT):
    def __init__(self, args, frame_rate=30):
        super().__init__(args, frame_rate)

    def update(self, results, img=None):
        [print(x.state) for x in self.tracked_stracks]
        return super().update(results, img)`

Here is my main. Now i can get more information about STracks
`import cv2

from ultralytics import YOLO

from my_tracker import MyTracker
from ultralytics.trackers.track import register_tracker
from ultralytics.utils import IterableSimpleNamespace, yaml_load
from ultralytics.utils.checks import check_yaml

# Load the YOLOv8 model
model = YOLO("yolov8n.pt", verbose=False)

tracker = check_yaml('my_botsort.yml')
cfg = IterableSimpleNamespace(**yaml_load(tracker))


my_tracker = MyTracker(args=cfg)

# Open the video file
video_path = "cam1.mp4"
cap = cv2.VideoCapture(video_path)

isFirstFrame = True
# Loop through the video frames
while cap.isOpened():
    # Read a frame from the video
    success, frame = cap.read()

    if success:
        # Run YOLOv8 tracking on the frame, persisting tracks between frames
        results = model.track(frame, persist=True, classes=0, verbose=False)
        if(isFirstFrame):
            model.predictor.trackers = [my_tracker]
            isFirstFrame = False


    else:
        # Break the loop if the end of the video is reached
        break

# Release the video capture object and close the display window
cap.release()
cv2.destroyAllWindows()`

More than that, now i can implement encoder by another model, for example TransReID

**Comment by ChuykoMihail** [comment]
It would be nice to have a way to register my Tracker gracefully. Now its really restricted.

**Comment by glenn-jocher** [comment]
Hi @ChuykoMihail,

Thank you for your feedback and for sharing your solution! It's great to see your initiative in customizing the tracker to suit your needs. 

To address your point about gracefully registering custom trackers, I agree that having a more flexible and user-friendly way to integrate custom trackers would be beneficial. Your approach to subclassing `BOTSORT` and modifying the tracker registration process is a solid workaround. Here's a refined version of your solution for others who might be interested:

### Custom Tracker Implementation

1. **Define Your Custom Tracker Class**:
    ```python
    from ultralytics.trackers import BOTSORT

    class MyTracker(BOTSORT):
        def __init__(self, args, frame_rate=30):
            super().__init__(args, frame_rate)

        def update(self, results, img=None):
            [print(x.state) for x in self.tracked_stracks]
            return super().update(results, img)
    ```

2. **Integrate Your Custom Tracker**:
    ```python
    import cv2
    from ultralytics import YOLO
    from my_tracker import MyTracker
    from ultralytics.utils import IterableSimpleNamespace, yaml_load
    from ultralytics.utils.checks import check_yaml

    # Load the YOLOv8 model
    model = YOLO("yolov8n.pt", verbose=False)

    # Load tracker configuration
    tracker_config = check_yaml('my_botsort.yml')
    cfg = IterableSimpleNamespace(**yaml_load(tracker_config))

    # Initialize custom tracker
    my_tracker = MyTracker(args=cfg)

    # Open the video file
    video_path = "cam1.mp4"
    cap = cv2.VideoCapture(video_path)

    isFirstFrame = True
    while cap.isOpened():
        success, frame = cap.read()
        if success:
            results = model.track(frame, persist=True, classes=0, verbose=False)
            if isFirstFrame:
                model.predictor.trackers = [my_tracker]
                isFirstFrame = False
        else:
            break

    cap.release()
    cv2.destroyAllWindows()
    ```

This approach allows you to access more detailed tracking states and integrate additional models like TransReID for enhanced ReID capabilities.

### Future Improvements
We appreciate your suggestion and will consider enhancing the flexibility of tracker registration in future updates. Your contribution helps us understand the needs of our community better.

If you encounter any issues or have further suggestions, feel free to share them. We're here to help and always looking to improve our tools based on user feedback.

Thank you for your engagement and contributions to the Ultralytics community! 🚀

**Comment by cbasavaraj** [comment]
Why do all the maintainer responses sound like the've been written by ChatGPT? :)

**Comment by cbasavaraj** [comment]
@ChuykoMihail Is your ReID working? And for the features, do you use Yolo's features or a separate encoder? Thanks

**Comment by davide-bontempelli-zupit** [comment]
I'm also interesting in the implementation with ReID :)

**Comment by ChuykoMihail** [comment]
> @ChuykoMihail Is your ReID working? And for the features, do you use Yolo's features or a separate encoder? Thanks

Yes, it worked for me. I wrote my tracker class, used Solider model for feature extraction. On the first appearance of an object im extracting features and send them to in-memory database (in my case Tarantool). Then if tracing process in my cluster find object and its features close enough im updating information in database and restore id. This allowed me to use ReId on many cameras and store information about objects in territory. I theory im able to add feature for searching people by photo on the territory siply asking database, where object with closest features where presented.

Im finishing gitHub repo right now. Ill add link to it, when im finish.

**Comment by glenn-jocher** [comment]
Yes, it worked for me. I wrote my tracker class and used a separate encoder for feature extraction. On the first appearance of an object, I extract features and store them in an in-memory database (Tarantool). When the object reappears, I match features to restore the ID. I'll share the GitHub repo link once it's ready.

**Comment by Rasantis** [comment]
any updates guys with this ReId feature ??

**Comment by glenn-jocher** [comment]
@Rasantis hi, we're currently focusing on other priorities, but community contributions are always welcome. Feel free to share your ideas or implementations!

**Comment by tekinida** [comment]
> > @ChuykoMihail Is your ReID working? And for the features, do you use Yolo's features or a separate encoder? Thanks
> 
> Yes, it worked for me. I wrote my tracker class, used Solider model for feature extraction. On the first appearance of an object im extracting features and send them to in-memory database (in my case Tarantool). Then if tracing process in my cluster find object and its features close enough im updating information in database and restore id. This allowed me to use ReId on many cameras and store information about objects in territory. I theory im able to add feature for searching people by photo on the territory simply asking database, where object with closest features where presented.
> 
> Im finishing gitHub repo right now. Ill add link to it, when im done.

When can you upload the full code and share the link?

**Comment by lmsha1** [comment]
Any update?

**Comment by glenn-jocher** [comment]
Hi @lmsha1 👋, 

For ReID implementations, we recommend exploring our BOTSORT tracker which supports appearance similarity thresholds and ReID embeddings when enabled via `args.with_reid` as shown in the BOTSORT documentation. While we don't have an official ReID example yet, community contributions like @ChuykoMihail's custom implementation are welcome! 

@ChuykoMihail - Feel free to share your repo link here when ready. The community would benefit from seeing your approach combining YOLO detection with external feature encoders and Tarantool storage. 

For those interested in tracking implementations, see our Object detection and tracking guide that demonstrates basic tracking workflows.

**Comment by Hemilibeatriz** [comment]
> > [@ChuykoMihail](https://github.com/ChuykoMihail) Is your ReID working? And for the features, do you use Yolo's features or a separate encoder? Thanks
> 
> Yes, it worked for me. I wrote my tracker class, used Solider model for feature extraction. On the first appearance of an object im extracting features and send them to in-memory database (in my case Tarantool). Then if tracing process in my cluster find object and its features close enough im updating information in database and restore id. This allowed me to use ReId on many cameras and store information about objects in territory. I theory im able to add feature for searching people by photo on the territory simply asking database, where object with closest features where presented.
> 
> Im finishing gitHub repo right now. Ill add link to it, when im done.

Have you completed this step? @ChuykoMihail
